samples/AvatarWithRoundedCorner/Program.cs

```diff
@@ -8,62 +8,74 @@ namespace AvatarWithRoundedCorner
     using SixLabors.Primitives;
     using SixLabors.Shapes;
 
-    class Program
+    static class Program
     {
         static void Main(string[] args)
         {
             System.IO.Directory.CreateDirectory("output");
+            using (var img = Image.Load("fb.jpg"))
+            {
+                // as generate returns a new IImage make sure we dispose of it
+                using (Image<Rgba32> dest = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 20)))
+                {
+                    dest.Save("output/fb.png");
+                }
 
-            GenerateAvatar("fb.jpg", "output/fb.png", new Size(200, 200), 20);
-            GenerateAvatar("fb.jpg", "output/fb-round.png", new Size(200, 200), 100);
-            GenerateAvatar("fb.jpg", "output/fb-rounder.png", new Size(200, 200), 150);
+                using (Image<Rgba32> destRound = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 100)))
+                {
+                    destRound.Save("output/fb-round.png");
+                }
+
+                using (Image<Rgba32> destRound = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 150)))
+                {
+                    destRound.Save("output/fb-rounder.png");
+                }
+                
+                // the original `img` object has not been altered at all.
+            }
         }
 
-        private static void GenerateAvatar(string source, string destination, Size size, float cornerRadius)
+        // lets create our custom image mutating pipeline
+        private static IImageProcessingContext<Rgba32> ConvertToAvatar(this IImageProcessingContext<Rgba32> operations, Size size, float cornerRadius)
         {
-            using (var image = Image.Load(source))
+            return operations.Resize(new ImageSharp.Processing.ResizeOptions
             {
-                image.Resize(new ImageSharp.Processing.ResizeOptions
-                {
-                    Size = size,
-                    Mode = ImageSharp.Processing.ResizeMode.Crop
-                });
-
-                ApplyRoundedCourners(image, cornerRadius);
-                image.Save(destination);
-            }
+                Size = size,
+                Mode = ImageSharp.Processing.ResizeMode.Crop
+            }).Apply(i => ApplyRoundedCorners(i, cornerRadius));
```

Comment:
@antonfirsov I think the difference is that we were looking at 200+ changes and have visibility of the internals which complicates things. There's a phrase in English "Can't see the wood for the trees" which describes perfectly my issues with understanding here.

We should flesh out and heavily comment our examples to make things easier for consumers to understand. I like the name `ApplyDelegateProcessor()` as it describes exactly what we are doing but do we need the user to be aware of the underlying operation? Maybe with the correct intellisense docs `Apply()` can be enough.